app/assets/javascripts/discourse/widgets/quick-access-panel.js.es6

```diff
@@ -0,0 +1,146 @@
+import { h } from "virtual-dom";
+
+import { createWidget } from "discourse/widgets/widget";
+import { headerHeight } from "discourse/components/site-header";
+
+const AVERAGE_ITEM_HEIGHT = 55;
+
+/**
+ * This tries to enforce a consistent flow of fetching, caching, refreshing,
+ * and rendering for "quick access items".
+ *
+ * There are parts to introducing a new quick access panel:
+ * 1. A user menu link that sends a `quickAccess` action, with a unique `type`.
+ * 2. A `quick-access-${type}` widget, extended from `quick-access-panel`.
+ */
+export default createWidget("quick-access-panel", {
+  tagName: "div.quick-access-panel",
+  emptyStatePlaceholderItemKey: "",
+
+  buildKey: () => {
+    throw Error('Cannot attach abstract widget "quick-access-panel".');
+  },
+
+  markReadRequest() {
+    return Ember.RSVP.Promise.resolve();
+  },
+
+  hasUnread() {
+    return false;
+  },
+
+  showAll() {},
+
+  hasMore() {
+    return this.state.items.length >= this.estimateItemLimit();
+  },
+
+  findStaleItems() {
+    return [];
+  },
+
+  findNewItems() {
+    return Ember.RSVP.Promise.resolve([]);
+  },
+
+  newItemsLoaded() {},
+
+  itemHtml(item) {}, // eslint-disable-line no-unused-vars
+
+  emptyStatePlaceholderItem() {
+    if (this.emptyStatePlaceholderItemKey) {
+      return h("li.read", I18n.t(this.emptyStatePlaceholderItemKey));
+    } else {
+      return "";
+    }
+  },
+
+  defaultState() {
+    return { items: [], loading: false, loaded: false };
+  },
+
+  markRead() {
+    this.markReadRequest().then(() => {
```

Comment:
It's a good habit to always return the promise in case something eventually wants to wait on it.